internal/exec/stages/files/units.go

```diff
@@ -64,10 +92,69 @@ func (s *stage) createUnits(config types.Config) error {
 			s.relabel(relabelpath)
 		}
 	}
-	// and relabel the preset file itself if we enabled/disabled something
-	if enabledOneUnit {
-		s.relabel(util.PresetPath)
+	// if we have presets then create the systemd preset file.
+	if len(presets) != 0 {
+		if err := s.createSystemdPresetFile(presets); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// parseInstanceUnit is a helper function to extract the name
+// and coresponding instances for a given instantiated unit.
+func parseInstanceUnit(unit types.Unit) (string, string, error) {
+	at := strings.Index(unit.Name, "@")
+	dot := strings.LastIndex(unit.Name, ".")
+	if dot == -1 {
+		return "", "", errors.ErrNoDotSuffix
+	}
+	instance := unit.Name[at+1 : dot]
+	serviceInstance := unit.Name[0:at+1] + unit.Name[dot:len(unit.Name)]
+	return serviceInstance, instance, nil
+}
+
+// createSystemdPresetFile creates the presetfile for enabled/disabled
+// instantiated/systemd units.
+func (s *stage) createSystemdPresetFile(presets map[string]*Preset) error {
+	hasInstanceUnit := false
+	for _, value := range presets {
+		if value.instantiateable {
+			hasInstanceUnit = true
+			break
+		}
+	}
+	for _, value := range presets {
+		unitString := value.unit
+		if value.instantiateable {
+			unitString = fmt.Sprintf("%s %s", unitString, strings.Join(value.instances, " "))
+		}
+		if value.enabled {
+			if err := s.Logger.LogOp(
+				func() error { return s.EnableUnit(unitString) },
+				"enabling unit %q", unitString,
```

Comment:
These log messages are going to look kind of strange with instantiated units, particularly with many instances.

Reply:
Can we do something to improve this?

Reply:
we could change it to `setting preset to {enabled|disabled} for "$unitstring"`

- setting preset to enabled for "sshd.service"
- setting preset to disabled for "echo@.service foo bar" 

WDYT?